Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -2722,6 +2728,148 @@ extension BrowserViewController: ContextMenuHelperDelegate {
     }
 }
 
+/**
+ A third party search engine Browser extension
+**/
+extension BrowserViewController {
+
+    /**
+    This allows us to find the part of the WKWebview that has access to
+    the inputAcessoryView
+    **/
+    func findWebViewFirstResponder(view:UIView) -> UIView? {
+        let subviews = view.subviews
+        if subviews.count == 0 {
+            return nil
+        }
+        for subview: UIView in subviews {
+            if subview.isFirstResponder() {
+                return subview
+            }
+            return findWebViewFirstResponder(subview)
+        }
+        return nil
+    }
+
+    func addCustomSearchButtonToWebView(webView:WKWebView) {
+        if self.customSearchEngineButton == nil {
+            self.customSearchEngineButton = self.searchEngineButton()
+        }
+
+        //check if the search engine has already been added.
+        let domain = webView.URL?.domainURL().host
+        let matches = self.profile.searchEngines.orderedEngines.filter({$0.shortName == domain})
+        if (!matches.isEmpty) {
+            self.customSearchEngineButton!.tintColor = UIColor.grayColor()
+            self.customSearchEngineButton?.userInteractionEnabled = false
+        } else {
+            self.customSearchEngineButton!.tintColor = UIConstants.SystemBlueColor
+            self.customSearchEngineButton?.userInteractionEnabled = true
+        }
+
+        guard let webContentView = findWebViewFirstResponder(webView),
+            let input = webContentView.performSelector(Selector("inputAccessoryView")),
+            let inputView = input.takeUnretainedValue() as? UIInputView,
+            let nextButton = inputView.valueForKey("_nextItem") as?  UIBarButtonItem,
+            let nextButtonView = nextButton.valueForKey("view") as? UIView else {
+                //failed to find the inputView instead lets use the findInPage bar
+                addCustomSearchButtonToFindInPage()
+                return
+            }
+
+            inputView.addSubview(self.customSearchEngineButton!)
+            self.customSearchEngineButton!.snp_remakeConstraints { make in
```

Comment:
I think these constraints can go into the `lazy var` block mentioned above. One of the nice properties of constraints is that you can set them once and be done!